Create number formatting options on the profile panel

joshmcrty opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change

<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue or discussion
  in the additional information section.
-->

This adds a Number Format option to the Profile panel that allows users to select how numbers are formatted in the UI. The available options are Auto (use language), comma for thousands separator, decimal for thousands separator, and none which bypasses the formatting function wherever it is used.

![Screen Shot 2020-12-07 at 11 26 11 AM](https://user-images.githubusercontent.com/1522068/101389674-06bd3f80-387f-11eb-813d-6638b4132037.png)

This is implemented by saving the format choice to the core user data. I'm not sure if any core/python changes are required; this seems to work fine in my dev environment with frontend changes only.

This PR updates the `computeStateDisplay` function to use a new argument -- `hass.userData` -- and updates all areas of the code that call this function to pass in the new argument. I opted for this approach so that in the future if other formatting features are created (for example date format options), they can be saved to the core user data and will already be available to `computeStateDisplay` without updating all uses of `computeStateDisplay` again.

The `formatNumber` function is also updated with new expected arguments to allow for format options. All direct uses of the `formatNumber` function have been updated.

## Type of change

<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR.
-->

Go to the profile page and select a number format. Lovelace dashboards should now display numbers that reflect the selected number format.

## Additional information

<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion: #7516, #7763, #7868, #7745 
- Link to documentation pull request:

## Checklist

<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

<!--
  Thank you for contributing <3
-->

[docs-repository]: https://github.com/home-assistant/home-assistant.io

joshmcrty committed:
Add number format options to profile panel

joshmcrty committed:
Update frontend to use number format options.

joshmcrty committed:
Localize number format options on profile page

kukulich commented:
I'm not sure if the two options (comma for thousands separator, decimal for thousands separator) make sense. Eg. in Czech we have space for thousands separator.

I'm happy user of the auto option (and I thank you very much for the implementation) but I can imagine some users may want to use English language and the Czech number formatting - just like other users for other languages in #7745

So I think the best implementation would be to have two options for decimal point and thousands separator.

joshmcrty commented:
@kukulich thank you for the feedback. I think you are right and I want to enable this feature to handle as many use-cases as possible. I'm going to rework the options a bit, because the current approach starts to get brittle with more and more options. Unfortunately there isn't a good browser localization API where I can arbitrarily specify what to use for the thousands separator and the decimal separator; it has to be a [Unicode BCP 47 locale identifier](https://www.unicode.org/reports/tr35/tr35.html#BCP_47_Conformance).

joshmcrty committed:
Handle space separator for thousands, add fallback locales for formats

joshmcrty committed:
Catch error if locale is invalid and use browser default locale

This enables using the "TEST" language

joshmcrty committed:
Add system and space_comma format options

joshmcrty committed:
Show example formatted number in description

joshmcrty commented:
The latest commits add an option for spaces as thousands separators, and a "Use system locale" option that uses the system default locale. This option is probably what most users should choose when they select English as the language but prefer native locale formatting.

The commits also add fallback locales for the three specific formatting options in case the first locale is not available on the users browser/operating system. This should ensure that the formatting works as expected in most cases.

A "live" formatted number is now in the description to provide quick visual feedback so users can verify the chosen format is working as expected.

<img width="629" alt="Screen Shot 2020-12-07 at 6 28 30 PM" src="https://user-images.githubusercontent.com/1522068/101426337-05f5cf00-38ba-11eb-9c02-99f0dcf03f45.png">

joshmcrty committed:
Avoid lint error

bramkragten commented on src/panels/profile/ha-pick-number-format-row.ts:
```diff
@@ -0,0 +1,111 @@
+        </span>
+        <span slot="description">
+          ${this.hass.localize("ui.panel.profile.number_format.description")}
+          ${formatNumber(1234567.89, {
+            language: this.hass.language,
+            format: this.coreUserData?.numberFormat,
+          })}
```
Can we add this as a secondary line to the `auto` and `system` options instead of in the description?

joshmcrty replied:
Done!

bramkragten commented on src/panels/profile/ha-pick-number-format-row.ts:
```diff
@@ -0,0 +1,111 @@
+    });
+  }
+
+  static get styles(): CSSResult {
+    return css`
+      a {
+        color: var(--primary-color);
+      }
+    `;
+  }
+}
```
unused?

joshmcrty replied:
Removed

bramkragten commented:
I'm thinking about where we should store these settings, currently, we already store the language per user in the frontend user date under `language`, we now add this to `core`, I think these settings should be together as they are all about localization. I think we should pass the format functions 1 object with localization options.

spacegaier commented:
I think I agree with Bram that this would be cleaner and just a logical:
```json
{
    "version": 1,
    "key": "frontend.user_data_ca2aa4188cea409b8b281db15bec6072",
    "data": {
        "language": {
            "language": "en",
            "number_format": "de"
        }
    }
}
```

At runtime we then have this one object with all localization info, so we keep lean and clean interfaces 👍 .

joshmcrty commented:
> I think I agree with Bram that this would be cleaner and just a logical:
> 
> ```json
> {
>     "version": 1,
>     "key": "frontend.user_data_ca2aa4188cea409b8b281db15bec6072",
>     "data": {
>         "language": {
>             "language": "en",
>             "number_format": "de"
>         }
>     }
> }
> ```
> 
> At runtime we then have this one object with all localization info, so we keep lean and clean interfaces 👍 .

@bramkragten @spacegaier saving the data to the `language` object is easy enough, but how should I access it? `hass.language` is just the string `"en"` from the `language` property in this example. I saved the `numberFormat` to the core user data because that was accessible from the `hass.userData` object.

bramkragten commented:
Yeah, it would need some refactoring

joshmcrty committed:
Update state-card-display LitElement to use number format options

joshmcrty committed:
Removed unused CSS

joshmcrty committed:
Move number format examples into second line on options

joshmcrty committed:
Remove old state-card-display.js file

joshmcrty committed:
Remove unused imports

joshmcrty committed:
Save number_format to user data language object instead of core object

joshmcrty committed:
Refactor hass.language as an object with language & number_format props

joshmcrty committed:
Refactor format_date and format_time functions

joshmcrty committed:
Reset devcontainer file

joshmcrty committed:
Minor code comment fix

joshmcrty commented:
@bramkragten I rebased and force pushed after all of the import sorting in dev. The latest changes include a refactor of the use of `hass.language` as an object instead of a string as well as addressing the comments from your previous review.

bramkragten commented on src/components/ha-gauge.ts:
```diff
@@ -29,7 +29,7 @@ export class Gauge extends LitElement {
   @property({ type: Number }) public value = 0;
 
-  @property({ type: String }) public language = "";
+  @property() public language;
```
```suggestion
  @property() public language!;
```

joshmcrty replied:
Done!

bramkragten commented on src/state/translations-mixin.ts:
```diff
@@ -53,10 +64,21 @@ export default <T extends Constructor<HassBaseEl>>(superClass: T) =>
           // We just get language from backend, no need to save back
-          this._selectLanguage(language, false);
+          this._selectLanguage(
+            { ...this.hass!.language, language: language },
```
We do save the `FrontendTranslationData` to the backend, but we only use the language of it and not the number format settings?

bramkragten replied:
Oh we do on the above function...

bramkragten replied:
But that means we now fetch the user data twice, we should combine the 2

joshmcrty replied:
This has been combined into a single call to the backend.

bramkragten commented on src/state/translations-mixin.ts:
```diff
@@ -78,33 +100,41 @@ export default <T extends Constructor<HassBaseEl>>(superClass: T) =>
-      this._updateHass({ language, selectedLanguage: language });
+      this._updateHass({
+        language,
+        selectedLanguage: language.language,
```
So we now only save `language.language` to localstorage, I think we should store the entire `language` object, so all language settings can be overruled on device level?

joshmcrty replied:
Done!

joshmcrty committed:
Only load backend user translation preferences once

joshmcrty committed:
Fix decorator for language property

joshmcrty commented:
@bramkragten the latest commits should address all of the comments from your last review.

bramkragten commented on .devcontainer/devcontainer.json:
```diff
@@ -30,5 +30,8 @@
     },
     "files.trimTrailingWhitespace": true
-  }
+  },
+  "mounts": [
+    "source=/Users/joshua/projects/homeassistant-frontend,target=/workspaces/frontend,type=bind,consistency=cached"
+  ]
```
```suggestion
  }
```

joshmcrty replied:
Fixed

joshmcrty committed:
Save entire hass.language object to selectedLanguage

joshmcrty committed:
Fix up after rebase

joshmcrty commented:
@bramkragten rebase is complete and all tests are passing.

JonahKr commented:
Will this be accessible like the hass language parameter to lovelace cards?
Would be interesting to know for something like [this](https://github.com/custom-cards/custom-card-helpers/blob/master/src/format-number.ts)

joshmcrty commented:
> Will this be accessible like the hass language parameter to lovelace cards?
> Would be interesting to know for something like [this](https://github.com/custom-cards/custom-card-helpers/blob/master/src/format-number.ts)

@JonahKr  I have not done any custom lovelace card development, so I'm not familiar with what data or methods are available to custom cards. If the card is using the `hass` object, then this would potentially be a breaking change because any card expecting `hass.language` to be a string would need to be adapted to use `hass.language.language`, and could also use `hass.language.number_format`. The custom-card-helper function you referenced should probably be updated if/when this PR is merged to match the changes made to that function here.

@bramkragten do you know how this may impact custom lovelace cards?

bramkragten commented:
I was thinking about that the other day, this will be a breaking change for custom cards, something we should think about.

JonahKr commented:
What would speak against leaving the language attribute as is and just create the `number_format` attribute directly under the hass element?

joshmcrty commented:
My original PR did not affect the `hass.language` value, and I don't have a strong opinion on where the `number_format` value is saved other than I don't think we should add it directly to `hass`. A few options come to mind to get this PR ready to merge:

1. Revert the commits in this PR that affect `hass.language` so this does not break custom cards and save the `number_format` on `core` again.
2. Remove the option for choosing a number format on the profile page and update the current `formatNumber()` function to use the browser locale instead of the selected `language`. I think that would be a more sensible default now that I've seen some of the feedback and issues raised since I introduced number formatting. I believe this is also how dates are formatted in many areas of the UI currently, using `toLocaleString()`.
3. Create a new `hass.locale` property (or some other appropriate name) to store locale-specific information and leave `hass.language` for now, deprecate it and maybe log a warning via a getter. Then eventually remove it once most custom card authors have a change to update their cards? We could eventually offer profile options for date and time formatting that could be saved on `hass.locale` too. This lets us pass a single object around to all formatting functions which was the primary rationale for using `hass.language` to save `number_format`.

Something like this:

``` javascript
hass: {
  ...
  get language() {
    console.warn("hass.language is deprecated. Please use hass.locale.language instead.");
    return this.locale.language;
  },
  get selectedLanguage() {
    console.warn("hass.selectedLanguage is deprecated. Please use hass.locale.language instead.");
    return this.locale.language;
  }
  locale: {
    language: "en",
    number_format: "auto"
  }
  ...
}
```

@bramkragten what do you think?

JonahKr commented:
I think Option 3 is the best solution to this one... This keeps everything tidy without breaking half of the custom cards in a blink of a second. The deprecation time should be a min of 3 months then is this case.

bramkragten commented:
I like option 3, we don't even really have to remove it.

joshmcrty committed:
Use hass.locale to store language and number_format preferences

joshmcrty committed:
Add tests for formatNumber

joshmcrty committed:
Use system/browser locale number formatting by default

This matches the behavior of the default date/time formatting

joshmcrty commented:
I completed another rebase and made the changes listed in option 3 above. I also changed the "auto" number_format to use the system/browser locale instead of using the selected language in Home Assistant. There is now a "language" option that uses the selected language for number formatting. I'm not sure if that's the approach we want to take; let me know if I should revert that change.

joshmcrty commented:
Here is a screenshot of the latest version:

<img src="https://user-images.githubusercontent.com/1522068/111007260-aa6d2e80-834b-11eb-9205-44915a084d46.png" width="600" />

spacegaier commented:
> 
> 
> I completed another rebase and made the changes listed in option 3 above. I also changed the "auto" number_format to use the system/browser locale instead of using the selected language in Home Assistant. There is now a "language" option that uses the selected language for number formatting. I'm not sure if that's the approach we want to take; let me know if I should revert that change.

In my mind following the HA selected language as default/auto makes the most sense and I think that is what most users would want. Not sure what the use case for following the browser locale is, so if we offer that then I think it should not be the default.

joshmcrty committed:
Revert change to auto setting; use language as auto

joshmcrty committed:
Update format_number test

joshmcrty committed:
Update other tests

joshmcrty committed:
Update new code to use hass.locale

bramkragten commented on src/util/hass-translation.ts:
```diff
@@ -41,9 +40,8 @@ function findAvailableLanguage(language: string) {
 /**
  * Get user selected language from backend
  */
-export async function getUserLanguage(hass: HomeAssistant) {
-  const result = await fetchTranslationPreferences(hass);
-  const language = result ? result.language : null;
```
Why was this logic moved?

joshmcrty replied:
@bramkragten this was done to avoid fetching data twice as mentioned in https://github.com/home-assistant/frontend/pull/7925#pullrequestreview-579235645. At this point we can skip using `getUserLanguage()` and just use `findAvailableLanguage()` instead. The actual fetch from the backend needs to return the entire object with the `language` and `number_format` and is done in the `translations-mixin.ts` file now. `getUserLanguage()` was only used once in the codebase and does not seem to be needed as a distinct function. I'll push a new commit to make this change.

bramkragten replied:
But why not just add the `number_format` to the existing function `getUserLanguage` and rename that function to `getUserLocale`?

joshmcrty replied:
Done!

bramkragten commented:
Since we now made it backwards compatible, we don't need to rename every usage of `hass.language` in this PR, it would make the PR a lot easier to review if it would just contain the parts needed for the number formatting options.

bramkragten commented on src/components/entity/state-info.ts:
```diff
@@ -84,7 +84,7 @@ class StateInfo extends LitElement {
 
     const oldHass = changedProps.get("hass") as HomeAssistant | undefined;
-    if (!oldHass || oldHass.language !== this.hass.language) {
+    if (!oldHass || oldHass.locale.language !== this.hass.locale.language) {
```
We should also rerender when another local option changes
```suggestion
    if (!oldHass || oldHass.locale !== this.hass.locale) {
```

bramkragten commented on src/components/ha-climate-state.ts:
```diff
@@ -97,17 +97,17 @@ class HaClimateState extends LitElement {
-        this.hass!.language
-      )} %`;
+        this.hass.locale
+      )}%`;
```
```suggestion
      )} %`;
```

bramkragten commented on src/components/ha-date-range-picker.ts:
```diff
@@ -43,7 +43,7 @@ export class HaDateRangePicker extends LitElement {
     if (changedProps.has("hass")) {
       const oldHass = changedProps.get("hass") as HomeAssistant | undefined;
-      if (!oldHass || oldHass.language !== this.hass.language) {
+      if (!oldHass || oldHass.locale.language !== this.hass.locale.language) {
```
```suggestion
      if (!oldHass || oldHass.locale !== this.hass.locale) {
```

bramkragten commented on src/components/ha-sidebar.ts:
```diff
@@ -245,7 +245,7 @@ class HaSidebar extends LitElement {
       hass.user !== oldHass.user ||
       hass.localize !== oldHass.localize ||
-      hass.language !== oldHass.language ||
+      hass.locale.language !== oldHass.locale.language ||
```
```suggestion
      hass.locale !== oldHass.locale ||
```

bramkragten commented on src/components/ha-sidebar.ts:
```diff
@@ -281,7 +281,7 @@ class HaSidebar extends LitElement {
 
     const oldHass = changedProps.get("hass") as HomeAssistant | undefined;
-    if (!oldHass || oldHass.language !== this.hass.language) {
+    if (!oldHass || oldHass.locale.language !== this.hass.locale.language) {
```
```suggestion
    if (!oldHass || oldHass.locale !== this.hass.locale) {
```

bramkragten commented on src/dialogs/more-info/controls/more-info-weather.ts:
```diff
@@ -68,7 +68,7 @@ class MoreInfoWeather extends LitElement {
     if (
       !oldHass ||
-      oldHass.language !== this.hass.language ||
+      oldHass.locale.language !== this.hass.locale.language ||
```
```suggestion
      oldHass.locale !== this.hass.locale ||
```

bramkragten commented on src/dialogs/notifications/persistent-notification-item.ts:
```diff
@@ -92,7 +92,7 @@ export class HuiPersistentNotificationItem extends LitElement {
 
     const d = new Date(notification.created_at!);
-    return d.toLocaleDateString(hass.language, {
+    return d.toLocaleDateString(hass.locale.language, {
```
Can we use format date here?

bramkragten commented on src/panels/config/logs/util.ts:
```diff
@@ -1,7 +1,11 @@
-export const formatSystemLogTime = (date, language: string) => {
+export const formatSystemLogTime = (
+  date,
+  language: FrontendTranslationData
```
```suggestion
  locale: FrontendTranslationData
```

bramkragten commented on src/panels/config/script/ha-script-picker.ts:
```diff
@@ -135,7 +135,7 @@ class HaScriptPicker extends LitElement {
         .route=${this.route}
         .tabs=${configSections.automation}
-        .columns=${this._columns(this.hass.language)}
+        .columns=${this._columns(this.hass.locale.language)}
```
```suggestion
        .columns=${this._columns(this.hass.locale)}
```

bramkragten replied:
`locale` is used for the format of the last triggered date

joshmcrty replied:
Yes it is, but it gets that direction from `this.hass.locale` without needing to pass anything in. I reverted the change as part of using the backwards-compatible `hass.language` to simplify the PR review process. I don't think this needs to pass anything in to the `this._columns()` function?

bramkragten commented on src/panels/logbook/ha-logbook.ts:
```diff
@@ -58,15 +58,19 @@ class HaLogbook extends LitElement {
     const languageChanged =
-      oldHass === undefined || oldHass.language !== this.hass.language;
+      oldHass === undefined ||
+      oldHass.locale.language !== this.hass.locale.language;
```
```suggestion
      oldHass.locale !== this.hass.locale;
```

bramkragten commented on src/panels/lovelace/cards/hui-alarm-panel-card.ts:
```diff
@@ -132,7 +132,7 @@ class HuiAlarmPanelCard extends LitElement implements LovelaceCard {
       !oldHass ||
       oldHass.themes !== this.hass!.themes ||
-      oldHass.language !== this.hass!.language
+      oldHass.locale.language !== this.hass!.locale.language
```
```suggestion
      oldHass.locale !== this.hass!.locale
```

bramkragten commented on src/panels/lovelace/cards/hui-button-card.ts:
```diff
@@ -113,7 +113,7 @@ export class HuiButtonCard extends LitElement implements LovelaceCard {
       !oldHass ||
       oldHass.themes !== this.hass!.themes ||
-      oldHass.language !== this.hass!.language
+      oldHass.locale.language !== this.hass!.locale.language
```
```suggestion
      oldHass.locale !== this.hass!.locale
```

bramkragten commented on src/panels/lovelace/cards/hui-gauge-card.ts:
```diff
@@ -130,7 +130,7 @@ class HuiGaugeCard extends LitElement implements LovelaceCard {
           .max=${this._config.max!}
           .value=${stateObj.state}
-          .language=${this.hass!.language}
+          .language=${this.hass!.locale.language}
```
```suggestion
          .locale=${this.hass!.locale}
```

bramkragten commented on src/components/ha-gauge.ts:
```diff
@@ -29,7 +30,7 @@ export class Gauge extends LitElement {
   @property({ type: Number }) public value = 0;
 
-  @property({ type: String }) public language = "";
+  @property() public language!: FrontendTranslationData;
```
```suggestion
  @property() public locale!: FrontendTranslationData;
```

bramkragten commented on src/panels/lovelace/cards/hui-glance-card.ts:
```diff
@@ -134,7 +134,7 @@ export class HuiGlanceCard extends LitElement implements LovelaceCard {
       !oldHass ||
       oldHass.themes !== this.hass!.themes ||
-      oldHass.language !== this.hass!.language
+      oldHass.locale.language !== this.hass!.locale.language
```
```suggestion
      oldHass.locale !== this.hass!.locale
```

bramkragten commented on src/panels/lovelace/cards/hui-logbook-card.ts:
```diff
@@ -101,7 +101,7 @@ export class HuiLogbookCard extends LitElement implements LovelaceCard {
       !oldHass ||
       oldHass.themes !== this.hass!.themes ||
-      oldHass.language !== this.hass!.language
+      oldHass.locale.language !== this.hass!.locale.language
```
```suggestion
      oldHass.locale !== this.hass!.locale
```

bramkragten commented on src/panels/lovelace/cards/hui-picture-glance-card.ts:
```diff
@@ -118,7 +118,7 @@ class HuiPictureGlanceCard extends LitElement implements LovelaceCard {
       !oldHass ||
       oldHass.themes !== this.hass!.themes ||
-      oldHass.language !== this.hass!.language
+      oldHass.locale.language !== this.hass!.locale.language
```
```suggestion
      oldHass.locale !== this.hass!.locale
```

bramkragten commented on src/panels/lovelace/cards/hui-weather-forecast-card.ts:
```diff
@@ -283,7 +283,7 @@ class HuiWeatherForecastCard extends LitElement implements LovelaceCard {
                         ${dayNight
                           ? html`
                               ${new Date(item.datetime).toLocaleDateString(
```
We should use `formatDate` here

joshmcrty replied:
`formatDate` does not support the options needed here to format the date as just the short weekday. I can submit a separate PR to add that capability and go through the codebase to use it where appropriate after this PR is merged.

bramkragten replied:
Could be the next `locale` option, date/time format 😅

joshmcrty replied:
Yes the thought has crossed my mind 😳

bramkragten commented on src/panels/lovelace/cards/hui-weather-forecast-card.ts:
```diff
@@ -298,12 +298,12 @@ class HuiWeatherForecastCard extends LitElement implements LovelaceCard {
                               )}
                             `
                           : html`
                               ${new Date(item.datetime).toLocaleDateString(
```
We should use `formatDate` here

joshmcrty replied:
`formatDate` does not support the options needed here to format the date as just the short weekday. I can submit a separate PR to add that capability and go through the codebase to use it where appropriate after this PR is merged.

bramkragten commented on src/panels/lovelace/common/has-changed.ts:
```diff
@@ -15,7 +15,7 @@ function hasConfigChanged(element: any, changedProps: PropertyValues): boolean {
     oldHass.connected !== element.hass!.connected ||
     oldHass.themes !== element.hass!.themes ||
-    oldHass.language !== element.hass!.language ||
+    oldHass.locale.language !== element.hass!.locale.language ||
```
```suggestion
    oldHass.locale !== element.hass!.locale ||
```

bramkragten commented on src/panels/lovelace/editor/card-editor/hui-card-picker.ts:
```diff
@@ -147,7 +147,7 @@ export class HuiCardPicker extends LitElement {
     }
 
-    if (oldHass.language !== this.hass!.language) {
+    if (oldHass.locale.language !== this.hass!.locale.language) {
```
```suggestion
    if (oldHass.locale !== this.hass!.locale) {
```

bramkragten commented on src/panels/lovelace/ha-panel-lovelace.ts:
```diff
@@ -68,7 +68,7 @@ class LovelacePanel extends LitElement {
       this.lovelace &&
       this.hass &&
-      this.lovelace.language !== this.hass.language
+      this.lovelace.language !== this.hass.locale.language
```
```suggestion
      this.lovelace.locale !== this.hass.locale
```

bramkragten commented on src/panels/lovelace/ha-panel-lovelace.ts:
```diff
@@ -285,7 +285,7 @@ class LovelacePanel extends LitElement {
       urlPath: this.urlPath,
       editMode: this.lovelace ? this.lovelace.editMode : false,
-      language: this.hass!.language,
+      language: this.hass!.locale.language,
```
```suggestion
      locale: this.hass!.locale,
```

bramkragten commented on src/panels/profile/ha-pick-language-row.js:
```diff
@@ -95,15 +95,17 @@ class HaPickLanguageRow extends LocalizeMixin(EventsMixin(PolymerElement)) {
-    if (newVal !== this.hass.language) {
-      this.fire("hass-language-select", { language: newVal });
+    if (newVal !== this.hass.locale.language) {
+      this.fire("hass-language-select", {
```
Should this now be `hass-locale-select`?

joshmcrty replied:
I kept it as-is for now because I created a separate event name for selecting the `number_format` that does not reload translations.

bramkragten commented on src/panels/profile/ha-pick-number-format-row.ts:
```diff
@@ -0,0 +1,124 @@
+          ${this.hass.localize("ui.panel.profile.number_format.description")}
+        </span>
+        <ha-paper-dropdown-menu
+          label="Number Format"
```
Use localize

bramkragten commented on src/panels/profile/ha-pick-number-format-row.ts:
```diff
@@ -0,0 +1,124 @@
+            @iron-select=${this._handleFormatSelection}
+            attr-for-selected="format"
+          >
+            <paper-item .format=${NumberFormat.language}>
```
You can iterate over the enum instead of hard coding it

bramkragten commented on src/state/translations-mixin.ts:
```diff
@@ -58,18 +69,33 @@ export default <T extends Constructor<HassBaseEl>>(superClass: T) =>
     }
 
-    protected hassConnected() {
+    protected async hassConnected() {
```
We can't just make this async, it is a protected function of the parent class.

bramkragten commented on src/state/translations-mixin.ts:
```diff
@@ -58,18 +69,33 @@ export default <T extends Constructor<HassBaseEl>>(superClass: T) =>
       if (this.hass?.panels && oldHass.panels !== this.hass.panels) {
-        this._loadFragmentTranslations(this.hass.language, this.hass.panelUrl);
+        this._loadFragmentTranslations(
+          this.hass.locale?.language,
```
Can `locale` be undefined? We init it with a value?

bramkragten commented on src/state/translations-mixin.ts:
```diff
@@ -58,18 +69,33 @@ export default <T extends Constructor<HassBaseEl>>(superClass: T) =>
-    protected hassConnected() {
+    protected async hassConnected() {
       super.hassConnected();
-      getUserLanguage(this.hass!).then((language) => {
-        if (language && this.hass!.language !== language) {
-          // We just get language from backend, no need to save back
-          this._selectLanguage(language, false);
-        }
-      });
```
We should keep this logic as it was, it should just return `FrontendTranslationData` instead of just the language.

bramkragten commented on src/state/translations-mixin.ts:
```diff
@@ -197,7 +231,7 @@ export default <T extends Constructor<HassBaseEl>>(superClass: T) =>
 
       // Ignore the repsonse if user switched languages before we got response
-      if (this.hass!.language !== language) {
+      if (this.hass!.locale?.language !== language) {
```
Can `locale` be undefined? We init it with a value?

bramkragten commented on src/state/translations-mixin.ts:
```diff
@@ -262,7 +296,7 @@ export default <T extends Constructor<HassBaseEl>>(superClass: T) =>
       await new Promise((resolve) => setTimeout(resolve, 0));
 
-      if (language !== (this.hass ?? this._pendingHass).language) {
+      if (language !== (this.hass ?? this._pendingHass).locale?.language) {
```
Can `locale` be undefined? We init it with a value?

bramkragten commented on src/state/translations-mixin.ts:
```diff
@@ -278,7 +312,7 @@ export default <T extends Constructor<HassBaseEl>>(superClass: T) =>
       };
 
-      if (language === (this.hass ?? this._pendingHass).language) {
+      if (language === (this.hass ?? this._pendingHass).locale?.language) {
```
Can `locale` be undefined? We init it with a value?

joshmcrty committed:
Remove unneeded function for getting available user language

joshmcrty committed:
Return empty string if format_number is called without a number value

joshmcrty committed:
Remove unused import

joshmcrty committed:
Fix space before percent symbol

bramkragten commented on src/state/translations-mixin.ts:
```diff
@@ -58,18 +69,35 @@ export default <T extends Constructor<HassBaseEl>>(superClass: T) =>
+          this.hass!.locale?.language !== userLanguage)
+      ) {
+        // We just get language and number_format from backend, no need to save back
+        this._selectLanguage(prefs, false);
```
We should not call `_selectLanguage` when just the `number_format` is changed, as that will reload all translations etc. and that is not needed ofc. Just when `language` is changed we should call `_selectLanguage`, if it is just `number_format` call `this._updateHass`.

joshmcrty replied:
Good point. I am now firing a separate event for selecting the `number_format` and a separate event listener that just updates `hass.locale`.

joshmcrty committed:
Use formatDateTime for persistent notification tooltip

joshmcrty committed:
Rerender when any locale option changes

joshmcrty committed:
Rename language to locale

joshmcrty committed:
Localize number format label

joshmcrty committed:
Revert to using hass.language if possible to simplify PR changes

joshmcrty committed:
Use separate event for selecting number_format

joshmcrty committed:
Rename language property to locale

joshmcrty committed:
Iterate over number formats to simplify template

joshmcrty committed:
Use actual formatted number instead of spaces in localized value

joshmcrty committed:
Revert to original logic for language; update function name

joshmcrty committed:
Ensure hass.language getter is retained when calling  _updateHass()

joshmcrty committed:
Remove unused code

joshmcrty committed:
Fix type errors

bramkragten commented on src/state/hass-base-mixin.ts:
```diff
@@ -48,6 +48,12 @@ export class HassBaseEl extends LitElement {
     }
-    this.hass = { ...this.hass, ...obj };
+
+    // Object.assign and the spread operator only copy static values from a getter to the new object; Object.defineProperty must be used to retain the getter for hass.language
+    this.hass = Object.defineProperty({ ...this.hass, ...obj }, "language", {
+      get() {
+        return this.locale.language;
+      },
+    });
```
We might want to reconsider using a getter, this is called a LOT, and I really don't want to make it any slower.

It would probably be better to update `language` whenever we update `locale`...

(also, we should not have the descriptor defined in 2 places, we should use `Object.getOwnPropertyDescriptors()`)

bramkragten commented on src/state/translations-mixin.ts:
```diff
@@ -14,10 +15,22 @@ import { storeState } from "../util/ha-pref-storage";
+  // for fire event
+  interface HASSDomEvents {
+    "hass-language-select": {
+      locale: FrontendTranslationData;
```
It should then only have the language, and not the entire locale.

bramkragten commented on src/state/translations-mixin.ts:
```diff
@@ -14,10 +15,22 @@ import { storeState } from "../util/ha-pref-storage";
+      locale: FrontendTranslationData;
+    };
+    "hass-number-format-select": {
+      locale: FrontendTranslationData;
```
And this only the number

bramkragten commented on src/state/translations-mixin.ts:
```diff
@@ -56,20 +72,30 @@ export default <T extends Constructor<HassBaseEl>>(superClass: T) =>
+        if (locale && this.hass!.language !== locale.language) {
+          // We just got language from backend, no need to save back
+          this._selectLanguage(locale, false);
+        } else if (
```
```suggestion
        }
        if (
```

bramkragten commented on src/state/translations-mixin.ts:
```diff
@@ -94,17 +120,35 @@ export default <T extends Constructor<HassBaseEl>>(superClass: T) =>
-      this._updateHass({ language, selectedLanguage: language });
+      this._updateHass({
+        locale,
+        selectedLanguage: locale.language,
```
So we should just update `language` here too.

bramkragten commented on src/state/translations-mixin.ts:
```diff
@@ -56,20 +72,30 @@ export default <T extends Constructor<HassBaseEl>>(superClass: T) =>
-          // We just get language from backend, no need to save back
-          this._selectLanguage(language, false);
+      getUserLocale(this.hass!).then((locale) => {
+        if (locale && this.hass!.language !== locale.language) {
```
```suggestion
        if (locale?.language && this.hass!.language !== locale.language) {
```

bramkragten commented on src/state/translations-mixin.ts:
```diff
@@ -56,20 +72,30 @@ export default <T extends Constructor<HassBaseEl>>(superClass: T) =>
+          // We just got language from backend, no need to save back
+          this._selectLanguage(locale, false);
+        } else if (
+          locale &&
```
```suggestion
          locale?.number_format &&
```

bramkragten commented on src/util/hass-translation.ts:
```diff
@@ -39,18 +42,27 @@ function findAvailableLanguage(language: string) {
   const result = await fetchTranslationPreferences(hass);
-  const language = result ? result.language : null;
-  if (language) {
+  const language = result?.language || hass.locale.language;
+  const number_format = result?.number_format || hass.locale.number_format;
```
```suggestion
  const language = result?.language;
  const number_format = result?.number_format;
```

joshmcrty replied:
@bramkragten these changes will result in type errors, because `getUserLocale()` returns `Promise<FrontendTranslationData>` and `language` and `number_format` could now be `undefined` which is not a valid value for either of those properties. I can update the `FrontendTranslationData` interface to allow for `null` or `undefined` values, is that how you'd prefer to handle the errors? Generally I avoid allowing those on type interfaces if possible, which is why I wrote in the fallback values.

The `hassConnected()` function checks for falsy values before called `_selectLanguage()` and `_selectNumberFormat()` so updating the interface should work to resolve the type errors.

bramkragten replied:
In that case, I would use a different interface here, or `Partial<FrontendTranslationData>`

bramkragten commented on src/util/hass-translation.ts:
```diff
@@ -39,18 +42,27 @@ function findAvailableLanguage(language: string) {
-  if (language) {
+  const language = result?.language || hass.locale.language;
+  const number_format = result?.number_format || hass.locale.number_format;
+  if (result?.language) {
```
```suggestion
  if (language) {
```

bramkragten commented on src/util/hass-translation.ts:
```diff
@@ -39,18 +42,27 @@ function findAvailableLanguage(language: string) {
   }
-  return null;
+  return {
+    language,
```
```suggestion
    language: null,
```

joshmcrty committed:
Pass individual property to fireEvent for language and number_format

joshmcrty committed:
Revert using getter for hass.language

joshmcrty committed:
Clean up logic

joshmcrty committed:
Return Partial<FrontendTranslationData> in getUserLocale

bramkragten commented on src/util/hass-translation.ts:
```diff
@@ -39,18 +42,27 @@ function findAvailableLanguage(language: string) {
   }
-  return null;
+  return {
+    language,
```
We should not return this language if we didn't find it in `availableLanguage`

bramkragten commented on src/state/connection-mixin.ts:
```diff
@@ -41,6 +42,10 @@ export const connectionMixin = <T extends Constructor<HassBaseEl>>(
         defaultPanel: DEFAULT_PANEL,
         language: getLocalLanguage(),
         selectedLanguage: null,
+        locale: {
+          language: getLocalLanguage(),
```
We could call `getLocalLanguage` once instead of twice here.

bramkragten commented on src/panels/profile/ha-pick-language-row.js:
```diff
@@ -96,13 +96,13 @@ class HaPickLanguageRow extends LocalizeMixin(EventsMixin(PolymerElement)) {
   ready() {
     super.ready();
-    if (this.hass && this.hass.language) {
+    if (this.hass && this.hass.locale && this.hass.language) {
       this.setLanguageSelection(this.hass.language);
```
```suggestion
    if (this.hass && this.hass.locale && this.hass.locale.language) {
      this.setLanguageSelection(this.hass.locale.language);
```

bramkragten commented on src/panels/config/automation/ha-automation-picker.ts:
```diff
@@ -83,7 +83,7 @@ class HaAutomationPicker extends LitElement {
                 ? formatDateTime(
                     new Date(automation.attributes.last_triggered),
-                    this.hass.language
+                    this.hass.locale
```
Because we use `this.hass.locale` here, we should use that as the memorize function parameter instead of `this.hass.language`.

joshmcrty committed:
Review cleanup

bramkragten commented:
:tada: Thanks @joshmcrty

JonahKr commented:
@iantrich just FYI... this will Require some changes for the `format number` method in Custom-Card-Helper and `localize` function at the  Boilerplate-Card after the next release
Will create some PRs if I have time for it

joshmcrty commented:
@bramkragten Thanks for your patience with all of the reviews, much appreciated!
